Sources/HTTPClient/HTTPClient.swift

```diff
@@ -0,0 +1,8 @@
+import Foundation
+import HTTPTypes
+
+public protocol HTTPClientProtocol {
+  associatedtype Body
+  associatedtype Data
```

Comment:
add associatedType Body, Data for `Data`, `ByteBuffer`, `[Int8]`.
https://github.com/apple/swift-http-types/issues/15#issuecomment-1646736121

Reply:
I think having associated types here is kinda defeating the purpose of having a general protocol level abstraction for an HTTP client. In my opinion, we should come up with standardised types for the type of data.

Furthermore, I see two more outstanding questions:
1. We should support optional trailers on both the request and response side
2. We should support bi-direction body streaming

Reply:
Should we adopt 1 or 2?

1. one associatedType(Data), data, body has same Type.
2. no associatedType, use Foundation.Data

Reply:
I don't think we should adopt any of the two. There is currently no "bag-of-bytes" currency type that is used across the ecosystems. Choosing either `Data` or `ByteBuffer` will come with downsides.

In my opinion, we have to solve this question in the standard library first before we can tackle providing a common API abstraction here.

Reply:
Can we adopt Foundation.Data without solving issue?
Because I think solving standard library is difficult, and take too time...

Apple/ swiftlang has any plan to solve it?

Reply:
> Can we adopt Foundation.Data without waiting for the issue to be resolved?

We can't adopt `Foundation.Data` across the ecosystem right now since the server ecosystem uses `ByteBuffer` and is built around that. Adopting `Foundation.Data` would incur unnecessary allocations at the API boundaries to bridge between the types.

> Does Apple/SwiftLang have any plans to resolve this?

There were previous discussions around this topic but there currently is no proposed solution. One of the focus areas of the Swift project right now is tackling ownership and lifetime in the standard library including in collections like `Array<UInt8>`. This might inform some of the next steps around a common bag-of-bytes type.

Sadly, I think this puts the proposal with creating a shared API abstraction for HTTP clients on hold right now. Until we can solve the bag-of-bytes problem at the language level. Now you can still create this protocol in your own code and conform the various clients to it.